File: packages/gavio-py/gavio/providers/anthropic.py
"""AnthropicAdapter — Messages API (Claude Sonnet, Haiku, Opus)."""

from __future__ import annotations

import os
import time

from ..exceptions import ConfigurationError, ProviderUnavailableError
from ..pricing import PricingProvider
from ..request import GavioRequest
from ..response import GavioResponse
from ..types import Message, TokenUsage
from ._http import post_json
from .base import ProviderAdapter

_DEFAULT_BASE_URL = "https://api.anthropic.com/v1"
_API_VERSION = "2023-06-01"


class AnthropicAdapter(ProviderAdapter):
    """Talks to the Anthropic Messages endpoint over HTTP.

    Anthropic splits the system prompt from the message list, so any
    ``role == "system"`` messages are extracted into the ``system`` field.
    """

    def __init__(
        self,
        api_key: str | None = None,
        base_url: str | None = None,
        timeout_seconds: float = 30.0,
        pricing: PricingProvider | None = None,
    ) -> None:
        super().__init__(pricing)
        self._api_key = api_key or os.environ.get("ANTHROPIC_API_KEY")
        self.base_url = (base_url or _DEFAULT_BASE_URL).rstrip("/")
        self.timeout_seconds = timeout_seconds

    @property
    def provider_name(self) -> str:
        return "anthropic"

    def _headers(self) -> dict[str, str]:
        if not self._api_key:
            raise ConfigurationError(
                "ANTHROPIC_API_KEY not set (pass api_key= or set the env var)"
            )
        return {
            "x-api-key": self._api_key,
            "anthropic-version": _API_VERSION,
        }

    @staticmethod
    def _split_system(messages: list[Message]) -> tuple[str | None, list[Message]]:
        system_parts = [m["content"] for m in messages if m.get("role") == "system"]
        chat = [m for m in messages if m.get("role") != "system"]
        system = "\n".join(system_parts) if system_parts else None
        return system, chat

    async def complete(self, request: GavioRequest) -> GavioResponse:
        started = time.monotonic()
        system, chat = self._split_system(request.messages)
        payload: dict = {
            "model": request.model,
            "messages": chat,
            "max_tokens": request.max_tokens,
            "temperature": request.temperature,
        }
        if system:
            payload["system"] = system

        data = await post_json(
            f"{self.base_url}/messages",
            payload,
            self._headers(),
            timeout=self.timeout_seconds,
        )
        content = "".join(
            block.get("text", "")
            for block in data.get("content", [])
            if block.get("type") == "text"
        )
        usage_data = data.get("usage", {})
        usage = TokenUsage(
            prompt_tokens=usage_data.get("input_tokens", 0),
            completion_tokens=usage_data.get("output_tokens", 0),
        )
        return self._build_response(
            request, content, usage, data.get("model", request.model), started
        )

    async def health_check(self) -> bool:
        try:
            self._headers()
            return True
        except ConfigurationError:
            return False
        except ProviderUnavailableError:
            return False

File: packages/gavio-js/src/providers/anthropic.ts
/** anthropicAdapter — Messages API (Claude Sonnet, Haiku, Opus). */

import { ConfigurationError } from '../errors.js'
import type { PricingProvider } from '../pricing.js'
import type { GavioRequest } from '../request.js'
import type { GavioResponse } from '../response.js'
import { TokenUsage } from '../types.js'
import type { Message } from '../types.js'
import { BaseProviderAdapter } from './base.js'
import { postJson } from './http.js'

const DEFAULT_BASE_URL = 'https://api.anthropic.com/v1'
const API_VERSION = '2023-06-01'

export interface AnthropicAdapterOptions {
  apiKey?: string
  baseUrl?: string
  timeoutMs?: number
  pricing?: PricingProvider
}

/**
 * Talks to the Anthropic Messages endpoint. Anthropic splits the system prompt
 * from the message list, so any `role === "system"` messages are extracted into
 * the `system` field.
 */
class AnthropicAdapter extends BaseProviderAdapter {
  private readonly apiKey: string | undefined
  private readonly baseUrl: string
  private readonly timeoutSeconds: number

  constructor(options: AnthropicAdapterOptions = {}) {
    super(options.pricing)
    this.apiKey = options.apiKey ?? process.env['ANTHROPIC_API_KEY']
    this.baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, '')
    this.timeoutSeconds = (options.timeoutMs ?? 30_000) / 1000
  }

  get providerName(): string {
    return 'anthropic'
  }

  private headers(): Record<string, string> {
    if (!this.apiKey) {
      throw new ConfigurationError(
        'ANTHROPIC_API_KEY not set (pass apiKey or set the env var)',
      )
    }
    return {
      'x-api-key': this.apiKey,
      'anthropic-version': API_VERSION,
    }
  }

  private static splitSystem(messages: Message[]): [string | null, Message[]] {
    const systemParts = messages
      .filter((m) => m.role === 'system')
      .map((m) => m.content)
    const chat = messages.filter((m) => m.role !== 'system')
    const system = systemParts.length > 0 ? systemParts.join('\n') : null
    return [system, chat]
  }

  async complete(request: GavioRequest): Promise<GavioResponse> {
    const started = performance.now()
    const [system, chat] = AnthropicAdapter.splitSystem(request.messages)
    const payload: Record<string, unknown> = {
      model: request.model,
      messages: chat,
      max_tokens: request.maxTokens,
      temperature: request.temperature,
    }
    if (system) payload['system'] = system

    const data = await postJson(
      `${this.baseUrl}/messages`,
      payload,
      this.headers(),
      this.timeoutSeconds,
    )
    const blocks = (data['content'] as Array<Record<string, unknown>>) ?? []
    const content = blocks
      .filter((b) => b['type'] === 'text')
      .map((b) => (b['text'] as string) ?? '')
      .join('')
    const usageData = (data['usage'] as Record<string, number>) ?? {}
    const usage = new TokenUsage(
      usageData['input_tokens'] ?? 0,
      usageData['output_tokens'] ?? 0,
    )
    return this.buildResponse(
      request,
      content,
      usage,
      (data['model'] as string) ?? request.model,
      started,
    )
  }

  async healthCheck(): Promise<boolean> {
    try {
      this.headers()
      return true
    } catch {
      return false
    }
  }
}

/** Factory: build an Anthropic provider adapter. */
export function anthropicAdapter(options: AnthropicAdapterOptions = {}): AnthropicAdapter {
  return new AnthropicAdapter(options)
}
